File: 2025/ruby/day01.rb
require_relative 'day'

class Day01 < Day # >

  # @example
  #   day.part1 #=> 3
  def part1
    turn_results = [50]
    dial = 50
    parse_input.each do |rotation|
      dial = (dial + rotation) % 100
      turn_results << dial
    end
    turn_results
      .keep_if { _1 == 0 }
      .count
  end

  # @example
  #   day.part2 #=> 6
  def part2
    dial = 50
    passes = 0

    parse_input.each do |rotation|
      passes += case
                when rotation.positive? # turning right is straightforward
                  (dial + rotation) / 100
                when dial > rotation.abs # turning left but the negative turn is smaller than the current dial
                  0
                when dial == 0 # turning left when the dial is already at zero
                  rotation.abs / 100
                else # turning left when the negative turn is larger than current dial (we'll pass 0 at least once)
                  1 + (dial + rotation).abs / 100
                end
      dial = (dial + rotation) % 100
      puts [dial, passes].inspect
    end

    passes
  end

  # @example
  #   parsed = day.parse_input
  #   parsed.first #=> -68
  #   parsed.last #=> -82
  #   parsed[-2] #=> 14
  def parse_input
    @input
      .split("\n")
      .map { |line|
        dir, clicks = line[0], line[1..].to_i
        clicks *= -1 if dir == "L" # left (toward lower numbers)
        clicks
      }
  end

  EXAMPLE_INPUT = File.read("../inputs/day01-example-input.txt")
end

File: 2025/go/day01/day01.go
package day01

import (
	"fmt"
	"log"
	"robbkidd/aoc2025/utils"
	"strconv"
	"strings"
)

func Run() {
	input, err := utils.ReadInput()
	if err != nil {
		log.Fatal("Couldn't read real input")
	}
	fmt.Printf("Part1: %s\n", Part1(input))
	fmt.Printf("Part2: %s\n", Part2(input))
}

func Part1(input string) string {
	rotations := parse(input)
	dial := 50
	zero_count := 0

	for _, rotation := range rotations {
		dial = (dial + rotation) % 100
		if dial == 0 {
			zero_count++
		}
	}
	return strconv.Itoa(zero_count)
}

func Part2(input string) string {
	rotations := parse(input)
	dial := 50
	zero_count := 0

	for _, rotation := range rotations {
		if rotation > 0 { // turning right is straightforward
			zero_count += (dial + rotation) / 100
		} else if dial > abs(rotation) { // turning left but the negative turn is smaller than the current dial
			continue
		} else if dial == 0 { // turning left when the dial is already at zero
			zero_count += abs(rotation) / 100
		} else { // turning left when the negative turn is larger than current dial (we'll pass 0 at least once)
			zero_count += 1 + abs(dial+rotation)/100
		}
		dial = (dial + rotation) % 100
		fmt.Printf("dial: %d, rotation: %d, zero count: %d\n", dial, rotation, zero_count)
	}
	return strconv.Itoa(zero_count)
}

func abs(n int) int {
	if n < 0 {
		return n * -1
	}
	return n
}

func parse(input string) []int {
	lines := strings.Split(input, "\n")
	rotations := make([]int, len(lines))

	for i, line := range lines {
		dir := line[0]
		num := line[1:]
		clicks, _ := strconv.Atoi(num)

		switch dir {
		case 'L':
			rotations[i] = clicks * -1
		case 'R':
			rotations[i] = clicks
		}
	}
	return rotations
}
